### ISSUE TITLE
PR testing instability

### ISSUE THREAD

**Issue by jwillenbring** [trilinos/Trilinos#3373]
@trilinos/framework

## Expectations

We need PR testing for Trilinos that runs reliably not only currently, but also when we continue to scale up the testing to include additional builds.

## Current Behavior

There are two primary causes of instability in the PR testing currently:

1) Communication issues
2) Machine overloading issues

We have seen a number of distinct types of communication issues. Here is a partial list:

a) clone/Fetch timeouts on GitHub
b) clone/fetch timeouts on internal gitlab-ex
c) failed reporting of results to CDash
d) Failure to communicate PR info from GitHub issues to autotester
e) Failure to communicate from autotester back to GitHub issues

Machine overloading issues have shown up in the form of internal compiler errors and test timeouts.

## Motivation and Context

We need to be able to run our PR testing in parallel (multiple PR testing at the same time and also multiple builds for each PR at the same time). If the environment won't scale, we will not be able to test PRs fast enough to make the system practical.

## Definition of Done

Communication errors occur on average less than one time per day. Machine overload issues occur less than twice a week.

## Possible Solution

 
We are looking into several possible solutions.

1) Reducing the -j for builds so we don't run into compiler errors
2) Reducing the number of available executors on testing nodes so that fewer tests can run at the same time
3) Speaking with networking staff about why operations are timing out
4) Reducing the number and/or frequency of PR testing instances to reduce the average communication traffic.

**Comment by lucbv** [comment]
@jwillenbring what is your recommendation for the current PR suffering from these instability?
Is there something we can do to limit the failures?

**Comment by jwillenbring** [comment]
> what is your recommendation for the current PR suffering from these instability?
Is there something we can do to limit the failures?

Given the types of failures, there really is not anything that can be done from the perspective of a person submitting PRs. The only thing a while back was that we were getting too many PRs started at the same time because there were so many retest attempts (which was understandable). We had to slow the flow of PRs to get things back to a slightly better state.

**Comment by jwillenbring** [comment]
Yesterday the @trilinos/framework team decided to mirror the Trilinos testing scripts repos on gitlab-ex to gitlab. This was done after @william76 set up a test to verify that the majority of communication failures were with these repos and that mirrors of those repos on gitlab did not suffer from the same problems. We still need to deal with overloaded machines and the timeouts on GitHub (we have had some success with retries for fetches from GitHub), but the source of failures that seemed to put us in a completely unsustainable state should be dealt with for now.

**Comment by jwillenbring** [comment]
I should also note that we are maintaining some test jobs so that hopefully the gitlab-ex admins can track down the problem eventually. We would like to be able to move back to the main repos at some point.

**Comment by prwolfe** [comment]
I started to open an issue around the communication errors to testing-vm and I think they make sense to just add here.

in PR #3460 @rppawlo stated

'''
@trilinos/framework - can someone help me get this PR merged? The autotester now attempted this 7 times over the last two days and each time a we get a random connection timeout failure from the autotester:

Error when uploading file: /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_intel_17.0.1/pull_request_test/Testing/20180919-1824/Configure.xml
Error message was: Failed to connect to testing-vm.sandia.gov port 80: Connection timed out
Problems when submitting via HTTP
configure submit error = -1
'''

The framework team discussed and reviewed past PR performance

'''
A raw output of failures by type
 
Pipeline Trilinos-PR-Clone-Test (Gitlab-EX)
* 29 Jobs
* 3 failures
* ascic166x3
 
Project Trilinos_pullrequest_gcc_4.8.4
* 29 jobs
* 20 failures / 10 from comm errors
* ascic142 to github x 3
* ascic166 to testing-vm x 6
* ascic142 to testing-vm
 
Project Trilinos_pullrequest_gcc_4.9.3
* 29 jobs
* 9 failures / 2 from comm errors
* ascic142 to github
* ascic157 to github
 
Project Trilinos_pullrequest_intel_17.0.1
* 29 jobs
* 11 failures / 4 from comm errors
* ascic158 to github x 2
* ascic166 to testing-vm x 2
 
Node assignments
·         Trilinos_pullrequest_intel_17.0.1 143,141,166,166,158x25
·         Trilinos_pullrequest_gcc_4.9.3 157x12,142x12,143x3,141,144x2
·         Trilinos_pullrequest_gcc_4.8.4 166x8,142x21
'''

It was also decided to change the order jobs run in to see if that affects the bandwidth or connectivity.

**Comment by jwillenbring** [comment]
The specific issues we were dealing with here have been resolved to the point that I will close this issue. We can open a new issue for any specific future problems.
